OpenDreamRuntime/Objects/MetaObjects/DreamMetaObjectDatum.cs

```diff
@@ -28,15 +30,68 @@ class DreamMetaObjectDatum : DreamMetaObjectRoot {
             dreamObject.SpawnProc("Del");
         }
 
-        public override DreamValue OnVariableGet(DreamObject dreamObject, string variableName, DreamValue variableValue) {
-            if (variableName == "type") {
-                return new DreamValue(dreamObject.ObjectDefinition.Type);
-            } else if (variableName == "parent_type") {
-                return new DreamValue(_dreamManager.ObjectTree.GetTreeEntry(dreamObject.ObjectDefinition.Type).ParentEntry.ObjectDefinition.Type);
-            } else if (variableName == "vars") {
-                return new DreamValue(DreamListVars.Create(dreamObject));
-            } else {
-                return base.OnVariableGet(dreamObject, variableName, variableValue);
+        public override DreamValue OnVariableGet(DreamObject dreamObject, string variableName, DreamValue variableValue)
+        {
+            return variableName switch
+            {
+                "type" => new DreamValue(dreamObject.ObjectDefinition.Type),
+                "parent_type" => new DreamValue(_dreamManager.ObjectTree.GetTreeEntry(dreamObject.ObjectDefinition.Type)
+                    .ParentEntry.ObjectDefinition.Type),
+                "vars" => new DreamValue(DreamListVars.Create(dreamObject)),
+                _ => base.OnVariableGet(dreamObject, variableName, variableValue)
+            };
+        }
+
+        public override void OnVariableSet(DreamObject dreamObject, string variableName, DreamValue variableValue,
+            DreamValue oldVariableValue)
+        {
+            base.OnVariableSet(dreamObject, variableName, variableValue, oldVariableValue);
+
+            if (variableName == "tag")
+            {
+                oldVariableValue.TryGetValueAsString(out var oldStr);
+                variableValue.TryGetValueAsString(out var tagStr);
+
+                // Even if we're setting it to the same string we still need to dequeue it
+                if (!string.IsNullOrEmpty(oldStr))
+                {
+                    var queue = _dreamManager.Tags[oldStr];
+                    if (queue.Count > 1)
+                    {
+                        if (queue.Peek() == dreamObject)
+                        {
+                            queue.Dequeue();
+                        }
+                        else
+                        {
+                            // We're working off of the assumption that the majority of tags are unique most of the time
+                            // But when it isn't we have to do this grossness to remove the correct obj from the queue
+                            var realList = queue.ToList();
+                            realList.Remove(dreamObject);
+                            var replacementQueue = new Queue<DreamObject>(realList);
+                            _dreamManager.Tags.Add(tagStr, replacementQueue);
```

Comment:
It may be better to replace the queues with `List<DreamObject>` to prevent from needing to do this. `Enqueue()` can be replaced with `Add()`, and `locate()` can return the first element in the list without it being any different.